### DISCUSSION TITLE
assertion error

### DISCUSSION THREAD

**Discussion by sahil-kaushik1** [Cyfrin/foundry-full-course-cu#2852] [General]
```function testmintandbalance( ) public {
        vm.prank(USER);
        basicnft.mintnft(dog);

        assert(basicnft.balanceOf(USER)==1);
        assert(keccak256(abi.encodePacked(dog)) == keccak256(abi.encodePacked(basicnft.tokenURI(0))));
        assert(USER==msg.sender);

    }```
    
    this is a test function 
    so on using the prank the msg.sender==USER  i guess?
    but the assertion is failing at last on comparing.
what could be the issue?

**Comment by EngrPips** [comment]
Hello @sahil-kaushik1, Do you know which of the assert is failing? Maybe rerun your test with the `-vvvvv` flag and share with us the stack trace of the test process

**Reply by sahil-kaushik1** [reply]
![image](https://github.com/user-attachments/assets/2765e7ef-6ffa-4f14-a619-a391757c8e42)
the last assertion

**Reply by EngrPips** [reply]
I see. Please use `vm.startBroadcast(USER)` instead of `vm.prank(USER)` and let see if the revert persist.

**Reply by sahil-kaushik1** [reply]
![image](https://github.com/user-attachments/assets/e370175b-05c0-4aac-bede-efde9254a4b7)
it aint worked

**Reply by heheboii11** [reply]
Please use ```assertEq(keccak256(abi.encodePacked(dog)), keccak256(abi.encodePacked(basicnft.tokenURI(0))));```
And let us know if you still facing the issue

**Reply by EngrPips** [reply]
@sahil-kaushik1 That means it is not the last assertion that is failing, please do one assert at a time to avoid confusion.

**Reply by sahil-kaushik1** [reply]
``` function testmintandbalance( ) public {
        
        vm.prank(USER);
        basicnft.mintnft(dog);
        assert(USER==msg.sender);
        
    }```
    
    
    simply imagine the above code .
my doubt is since we are doing prank(user) why the msg.sender is not same as user?

![image](https://github.com/user-attachments/assets/b9cefb17-ddda-4f56-98db-db52c95e41dc)

**Reply by EngrPips** [reply]
Okay, to make things clear `vm.prank(USER)` is only active for the next line that means the `msg.sender` for this line `basicnft.mintnft(dog);` will be `USER` but the `msg.sender` for this line `assert(USER==msg.sender);` will be the `DEFAULT_FOUNDRY_SENDER` I assume but if you have the code as below, then the `msg.sender` should be the `USER` 

```js
vm.startPrank(USER);
        basicnft.mintnft(dog);
        assert(USER==msg.sender);
vm.stopPrank();
```

**Reply by sahil-kaushik1** [reply]
nope still getting that assertion error
![image](https://github.com/user-attachments/assets/6c27410a-e98c-4762-a69f-f247fbd5015f)


btw can u also tell the diff between startbroadcast and startprank

**Reply by EngrPips** [reply]
```js
vm.startPrank(USER);
        basicnft.mintnft(dog);
        console.log(USER);
        console.log(msg.sender);
        assert(USER==msg.sender);
vm.stopPrank();
```
Please use the above and let's what will be printed to the log.

**Reply by EngrPips** [reply]
`startBroadcast` is meant to be used in a script to specify the private key that sent transaction, while `startPrank` is meant to be used in a test contract to specify the sender of a transaction. That is what makes sense to me, but maybe you can read the Foundry docs https://book.getfoundry.sh/ to find out what the actual difference is.

**Reply by sahil-kaushik1** [reply]
```
vm.startPrank(USER);
        basicnft.mintnft(dog);
        console.log(USER);
        console.log(msg.sender);
        assert(USER==msg.sender);
vm.stopPrank();
```

![image](https://github.com/user-attachments/assets/c551c58f-e74b-4fb2-9274-ad11715b007f)

**Reply by EngrPips** [reply]
I see that inside the test contract, `DefaultSender` remains the `msg.sender`. I think you should read the `vm.prank` and `vm.startPrank` sections of the Foundry Book https://book.getfoundry.sh/cheatcodes/prank to understand properly how these cheat codes work. I assume that whenever you use `vm.startPrank` or `vm.prank` Foundry will only switch the `msg.sender` to the specified address just before it processes the transaction and at every other time the `msg.sender` in the test contract will normally be the `DefaultSender` but I am just trying to make sense of the situation you should read the docs to better understand how stuff works.

**Reply by sahil-kaushik1** [reply]
oki btw thanks for the help .
i will update if found the solution to this.

**Reply by LordAdhaar** [reply]
Hi @sahil-kaushik1,

From my understanding, `msg.sender` is the `DefaultSender` since you are calling it from the test. In the context of the test, `msg.sender` is the `DefaultSender`.

If you were to put a `console.log(msg.sender)` inside the `basicnft.mint()` function, you would get the `msg.sender` as `USER` because the `msg.sender == USER` in the context of the `basicnft.mint()` function.

However, for the test, the `msg.sender == DefaultSender`.

**Reply by EngrPips** [reply]
Make sense
